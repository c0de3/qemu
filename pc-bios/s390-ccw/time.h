#ifndef TIME_H
#define TIME_H

static inline u64 get_clock(void)
{
    u64 r;

    asm volatile("stck %0" : "=Q" (r) : : "cc");
    return r;
}

static inline u64 get_time_ms(void)
{
    /* Bit 51 is incremented each microsecond */
    return (get_clock() >> 12) / 1000;
}

static inline u64 get_time_seconds(void)
{
    return (get_time_ms()) / 1000;
}

static inline void yield(void)
{
    asm volatile ("diag 0,0,0x44"
                  : :
                  : "memory", "cc");
}

static inline void sleep(unsigned int seconds)
{
    ulong target = get_time_seconds() + seconds;

    while (get_time_seconds() < target) {
        yield();
    }
}

#endif
