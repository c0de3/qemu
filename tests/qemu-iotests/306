#!/usr/bin/env python3
#
# Test nbd reconnect on open
#
# Copyright (c) 2020 Virtuozzo International GmbH
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

import time

import iotests
from iotests import qemu_img_create, file_path, qemu_io_popen, qemu_nbd

disk, nbd_sock = file_path('disk', 'nbd-sock')
qemu_img_create('-f', iotests.imgfmt, disk, '1M')

client = qemu_io_popen('--image-opts', '-c', 'read 0 1M',
                       'driver=nbd,__virtuozzo_open-reconnect-attempts=4,'
                       'server.type=unix,server.path=' + nbd_sock)

time.sleep(2)

qemu_nbd('-k', nbd_sock, '-f', iotests.imgfmt, disk)

iotests.log(client.communicate()[0], filters=[iotests.filter_qemu_io])

client = qemu_io_popen('--image-opts', '-c', 'read 0 1M',
                       'driver=nbd,__virtuozzo_open-reconnect-attempts=1,'
                       'server.type=unix,server.path=' + nbd_sock)

time.sleep(1.5)

iotests.log(client.communicate()[0], filters=[iotests.filter_qemu_io,
                                              iotests.filter_testfiles])
