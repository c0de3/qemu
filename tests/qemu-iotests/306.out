read 1048576/1048576 bytes at offset 0
1 MiB, X ops; XX:XX:XX.X (XXX YYY/sec and XXX ops/sec)

can't open: Failed to connect socket TEST_DIR/PID-nbd-sock: No such file or directory

